🚀 3단계 - 수강신청(DB 적용)

Perhona opened the pull request:
석홍님 안녕하세요.

3단계 PR 확인 부탁드립니다!

이번 3단계는 JdbcTemplate으로 작업하면서 다소 복잡하게 느껴졌던 부분이 많았던 것 같습니다.

그 중에서 CoverImage 저장, Session 정보 가져오기를 할 때 set을 회피하기 위해 레포지토리 메서드에 직접 sessionId와 coveerImage를 넘겨준 부분이 있는데, 적합한 방향인지 궁금합니다.

다른 부분도 자세히 살펴봐주시면 감사하겠습니다.

ps. 리뷰 기간이 얼마 남지는 않았지만, 4단계까지 다 끝내지 못해도 좋으니 최대한 꼼꼼히 짚고 넘어가고 싶습니다!

이번에도 잘 부탁드려요~ 😄

Perhona committed:
docs(step3): 3단계 요구사항 추가, 2단계 피드백 일부 반영

Perhona committed:
refactor(step2): 수강신청을 Enrollment 객체가 수행하도록 수정

- SessionTest 에 있던 테스트를 역할에 따라 재분배

Perhona committed:
docs(step3): schema.sql 테이블 추가

Perhona committed:
feat(step3): sesison 관련 repository 코드 추가

Perhona committed:
test(step3): repository test 추가

- sessionRepository에서 generatedKey를 가져오는 로직 수정

seokhong commented on src/main/java/nextstep/courses/infrastructure/JdbcCoverImageRepository.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    @Override
+    public int save(CoverImage coverImage, Long sessionId) {
```
> 그 중에서 CoverImage 저장, Session 정보 가져오기를 할 때 set을 회피하기 위해 레포지토리 메서드에 직접 sessionId와 coveerImage를 넘겨준 부분이 있는데, 적합한 방향인지 궁금합니다.

이 코드와 관련이 있는 질문이라고 생각되어, 여기에 제 생각 답변으로 남깁니다. :)

1. `CoverImage`가 항상 `Session`과 함께 관리되는 정보라면, `CoverImage`를 `Session`과 분리하지 않고, 하나의 테이블로 관리할 수도 있을 것 같다는 생각이 들었습니다. 1:1 관계로 보이는데, 테이블을 분리했던 이유부터 고민을 시작해보셔도 좋겠습니다.
2. 현재 구조를 유지한다는 전제하에, `CoverImage`는 항상 연결된 `Session`을 알고 있을 것(`cover_image` 테이블의 PK가 `session_id` 이므로)으로 보입니다. 그렇다면 `CoverImage`를 만들 때 `sessionId` 필드도 받아서 설정해주는 것은 어떨까요? 객체의 관점에서, `CoverImage`에게 `Session`이 필수 정보라면, 생성자에서 그 값을 받도록 강제하고, 항상 존재하는지 검증하는 것이 좋다고 생각합니다.
3. 코드를 보면 `-Repository` postfix를 사용하셨는데, Repository 와 DAO의 차이는 무엇일까요? 관련해서 [블로그](https://80000coding.oopy.io/dcd0f224-053f-456c-af22-d7e0946fa868) 글을 참고해보셔도 도움이 될 것 같습니다. 현재의 구현을 보면, (`CoverImageRepository`의 구현체로 Jdbc를 사용한다는 것을 고려하더라도) `Session`이라는 도메인에 대한 Repository 보다는 매핑된 테이블을 추상화한 DAO에 좀 더 가까운 형태의 설계인 것처럼 보이기도 합니다. :)

seokhong commented on src/main/java/nextstep/courses/domain/ImageShape.java:
```diff
@@ -16,6 +16,12 @@ public ImageShape(Long width, Long height) {
+		if (width == null || height == null) {
+			throw new IllegalArgumentException("이미지의 가로 세로 값은 빈 값이 올 수 없습니다.");
+		}
+		if (width == 0L || height == 0L) {
```
꼼꼼한 검증 좋습니다. 👍 
아래의 `MIN_WIDTH`, `MIN_HEIGHT` 검증을 통해서도 예외가 발생할 것으로 보이는데, 추가로 밸리데이션을 분리하신 이유도 있을까요?
가로, 세로 값이 필수라면 wrapper 타입 대신 primitive 타입을 사용하는 건 어떨까요?

Perhona replied:
클래스를 설계할 때 무의식적으로 Wrapper 타입을 사용하려 한 것 같습니다.
찾아보니 Wrapper타입은 크게 null값을 가질 수 있고, generic 타입으로 설정할 수 있다는 점이 눈에 띄는데요. 해당 경우 외에는 오히려 primitive 타입을 사용하는 것이 안정성 측면에서 좋을까요?

seokhongkim replied:
[블로그](https://konstantinmb.medium.com/understanding-primitive-types-and-wrapper-classes-in-java-a-comprehensive-guide-6013c6b1c87)를 참고해보시면, primitive type이 null safe 하다는 장점 외에도 성능이나 메모리 효율성 측면에서도 장점이 있습니다.

seokhong replied:
[블로그](https://konstantinmb.medium.com/understanding-primitive-types-and-wrapper-classes-in-java-a-comprehensive-guide-6013c6b1c87)를 참고해보시면, primitive type이 null safe 하다는 장점 외에도 성능이나 메모리 효율성 측면에서도 장점이 있습니다.

seokhong commented on src/main/java/nextstep/courses/domain/Enrollment.java:
```diff
@@ -0,0 +1,22 @@
+
+public class Enrollment {
+    private Session session;
+    private Long userNumber;
```
여기도 수강 인원은 항상 존재하는 값이니(0명도 0이라는 값이 존재), primitive 타입을 사용하는 건 어떨까요?

seokhong commented on src/main/java/nextstep/courses/domain/Enrollment.java:
```diff
@@ -0,0 +1,22 @@
+    }
+
+    public NsUserSession enroll(Payment payment) throws CannotEnrollException {
+        session.checkStatus();
```
`enroll` 과정에서 필요한 모든 검증을 하나의 메서드로 제공하는 건 어떨까요?
`session.enroll(payment, userNumber)`와 같이 강의에게 등록하라고 요청을 보낼 수도 있을 것 같아요. 그렇게 하려면 수강 인원을 강의가 관리해야 하니, `Session`에 `nsUserSessions` 필드가 있어도 좋겠네요.

Perhona replied:
> session.enroll(payment, userNumber)와 같이 강의에게 등록하라고 요청을 보낼 수도 있을 것 같아요. 그렇게 하려면 수강 인원을 강의가 관리해야 하니, Session에 nsUserSessions 필드가 있어도 좋겠네요.

인자로 넘긴 userNumber로 Session이 maxUser와 비교 후  강의 조건에 맞지 않는다면 수강신청이 되지 않는 시나리오로 생각했는데, Session nsUserSessions를 필드로 갖고 있어야 하는 부분이 이해가 잘 되지 않습니다. 어떤 뜻인지 더 설명해주실 수 있을까요?? 😮

seokhong replied:
지금처럼 수강신청 과정에서 검증해야 할 내용을 한 번의 메서드 호출로만 변경하셔도 괜찮습니다!

현재 `userNumber`를 만들기 위해 `List<NsUserSessiion>`을 인자로 받게 되는데요.
`Enrollment`를 만들지 않고, `Session`을 통해 직접 수강신청을 한다면, `Session`에서 `List<NusUserSession>`을 가지고 있어도 될거라고 생각했습니다. 이렇게 변경하면, `session.enroll(payment)` 또는 `session.enroll(nsUserSession, payment)`와  같이 요청하고 `Session`이 직접 수강 신청을 처리할 수도 있을 것 같습니다.

`Session`이 `List<NsUserSession>`을 가지고 있는 경우 이점은, 강의가 "수강 신청을 한 학생이 누구인지 안다"는 것이라고 생각합니다. 수강 신청한 사람이 누구인지 강의가 직접 판단할 수 있고, 현재 인원도 강의가 알고 있겠죠. 강의와 관련된 정보가 모두 강의에게 있으므로 응집도가 더 높다고 생각하고, 이 정보를 바탕으로 필요한 검증이나 제약사항도 강의 객체를 통해 모두 할 수 있을거라고 생각했습니다.

seokhong commented on src/main/java/nextstep/courses/domain/SessionPaymentCondition.java:
```diff
@@ -7,12 +7,20 @@ public class SessionPaymentCondition {
         this.maxUser = maxUser;
     }
 
     public void checkPaidSession(Payment payment, Long userNumber) throws CannotEnrollException {
```
이 메서드는 무슨 목적으로 사용하길 기대하고 제공한 것일까요? 그에 맞춰 이름을 지어보시면 좋겠습니다.
유료 강의인지 확인하는 목적보다는, 수강 신청이 가능한지 금액과 인원을 검증하는 기능을 제공하는 것으로 보이네요. :)

seokhong commented on src/main/java/nextstep/courses/infrastructure/JdbcCoverImageRepository.java:
```diff
@@ -0,0 +1,35 @@
+    @Override
+    public int save(CoverImage coverImage, Long sessionId) {
+        String sql = "insert into cover_image (session_id, size, image_type, width, height) values (?, ?, ?, ?, ?)";
+        return jdbcTemplate.update(sql, sessionId, coverImage.size(), coverImage.imageType().name(), coverImage.imageShape().width(), coverImage.imageShape().height());
```
여기서도 객체를 저장한 후 값을 반환한다면, 아이디를 반환해보는 건 어떨까요?
`update` 메서드의 기본 반환 값은 변경된 로우의 개수로 알고 있습니다. `save` 이후에 반환된 값이 저장된 행의 개수라고 생각하기는 어려울 것 같아요.

seokhong commented on src/main/java/nextstep/courses/service/SessionService.java:
```diff
@@ -1,17 +1,35 @@
+        coverImageRepository.save(coverImage, sessionId);
+    }
+
     public void enroll(Payment payment) throws CannotEnrollException {
```
이 메서드에는 `@Transactional`이 누락된 것으로 보이네요.
메서드에 트랜잭션 설정을 안 하는 실수 상황을 방지하기 위해, 클래스에도 `@Transactional` 어노테이션을 사용해보면 어떨까요?

seokhong commented on src/test/java/nextstep/courses/domain/EnrollmentTest.java:
```diff
@@ -0,0 +1,36 @@
+            fail(e.getMessage());
+        }
+    }
+}
```
테스트 파일의 마지막에 newline 이 없다고 깃헙에서 경고를 띄워주네요.
파일 마지막의 newline 은 무슨 역할을 하는지 학습해보신 후, IDE 설정을 통해 휴먼 에러를 방지해보시면 좋겠습니다.

seokhong reviewed (changes requested):
남은님, 3단계 미션 잘 구현하셨습니다. 👍 
질문 남겨주신 내용에 대한 의견 포함하여 코드에 코멘트 남겼습니다.
코멘트 확인 부탁드립니다. 🙂

Perhona committed:
refactor(step3): 수강신청 도메인 관련 피드백 사항 반영

- NsUserSessions 일급 컬렉션 추가
- SessionPaymentCondition 클래스의 클래스명, 메서드명 수정
- ImageShape의 멤버변수 타입 변경으로 관련 validate 제거

Perhona committed:
refactor(step3): DB 매핑 부분 피드백 사항 반영

- Session을 생성할 때 CoverImage와 분리
- Repository의 수행 기능이 DAO에 가까워 명칭 변경

Perhona commented:
석홍님 안녕하세요~ 😄 
피드백 주신 부분 반영해보았습니다.

1. Session과 CoverImage를 분리한 이유를 생각해보았습니다.
- 결론적으로는, CoverImage는 Session이 로직을 수행할 때 항상 필요한 정보는 아니라는 생각이 들어, 둘을 분리하고 Builder패턴을 활용해 필요 시에 set 해주도록 하였습니다.

2. DAO vs Repository
- 제가 이해한 것이 맞다면, 현재 설계상으로는 테이블을 추상화한 DAO에 가까운 것이 맞는 것 같습니다.
- 만약 Repository로 변경하고자 한다면, get 혹은 save 시에 DAO를 여러개 사용해서라도 서비스단에서 사용되는 도메인 객체와 비슷하게 만들어 활용하게 하는 역할이 맞을까요?

3. Session과 현재 수강인원
- Session이 현재 수강인원을 관리하는 부분이 맞는 것인지 의문이 들어 위에 댓글을 남겨두었습니다.

감사합니다! :star2:

seokhong reviewed (approved):
남은님, 3단계 미션 코멘트 잘 반영하셨습니다. 👍 
3단계 PR 머지하겠습니다.
코멘트 남긴 내용은 다음 단계 미션 진행하실 때 참고해보시면 좋겠습니다. 이제 마지막 단계 미션만 남았네요. 4단계 미션에서 뵙겠습니다! 🙂
